django/db/models/sql/compiler.py

```diff
@@ -1253,21 +1253,21 @@ def get_related_klass_infos(klass_info, related_klass_infos):
 
         if restricted:
             related_fields = [
-                (o.field, o.related_model)
+                (o, o.field, o.related_model)
                 for o in opts.related_objects
                 if o.field.unique and not o.many_to_many
             ]
-            for related_field, model in related_fields:
-                related_select_mask = select_mask.get(related_field) or {}
+            for related_object, related_field, model in related_fields:
                 if not select_related_descend(
                     related_field,
                     restricted,
                     requested,
-                    related_select_mask,
+                    select_mask,
```

Comment:
`related_field` must be checked againt `select_mask` and not `select_mask.get(related_field)` just like any field. See how it's done in the `for f in opts.fields` loop above.